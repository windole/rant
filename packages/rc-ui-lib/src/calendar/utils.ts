/* eslint-disable no-nested-ternary */
export function compareMonth(date1: Date, date2: Date): number {
  const year1 = date1.getFullYear();
  const year2 = date2.getFullYear();

  if (year1 === year2) {
    const month1 = date1.getMonth();
    const month2 = date2.getMonth();
    return month1 === month2 ? 0 : month1 > month2 ? 1 : -1;
  }

  return year1 > year2 ? 1 : -1;
}

export function compareDay(day1: Date, day2: Date): number {
  const compareMonthResult = compareMonth(day1, day2);

  if (compareMonthResult === 0) {
    const date1 = day1.getDate();
    const date2 = day2.getDate();
    return date1 === date2 ? 0 : date1 > date2 ? 1 : -1;
  }

  return compareMonthResult;
}

export const cloneDate = (date: Date): Date => new Date(date);

export const cloneDates = (dates: Date | Date[]): Date | Date[] =>
  Array.isArray(dates) ? dates.map(cloneDate) : cloneDate(dates);

export function getDayByOffset(date: Date, offset: number): Date {
  const cloned = cloneDate(date);
  cloned.setDate(cloned.getDate() + offset);
  return cloned;
}

export const getPrevDay = (date: Date): Date => getDayByOffset(date, -1);
export const getNextDay = (date: Date): Date => getDayByOffset(date, 1);
export const getToday = (): Date => {
  const today = new Date();
  today.setHours(0, 0, 0, 0);
  return today;
};

export function calcDateNum(date: [Date, Date]): number {
  const day1 = date[0].getTime();
  const day2 = date[1].getTime();
  return (day2 - day1) / (1000 * 60 * 60 * 24) + 1;
}

export const getMonthEndDay = (year: number, month: number): number =>
  32 - new Date(year, month - 1, 32).getDate();

const calendarLang = {
  end: '结束',
  start: '开始',
  title: '日期选择',
  confirm: '确定',
  startEnd: '开始/结束',
  weekdays: ['日', '一', '二', '三', '四', '五', '六'],
  monthTitle: (year: number, month: number) => `${year}年${month}月`,
  rangePrompt: (maxRange: number) => `最多选择 ${maxRange} 天`,
};

export const t = (key: string, ...args): string => {
  if (args.length) {
    return calendarLang[key](...args);
  }
  return calendarLang[key];
};

export const formatMonthTitle = (date: Date): string =>
  t('monthTitle', date.getFullYear(), date.getMonth() + 1);
