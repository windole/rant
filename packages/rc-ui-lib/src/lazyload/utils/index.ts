import { ListenEvent } from '../PropsType';

/* eslint-disable prefer-rest-params */
const inBrowser = typeof window !== 'undefined';

export const DEFAULT_URL =
  'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7';

// 需要监听的事件
export const DEFAULT_EVENTS: ListenEvent[] = [
  'scroll',
  'wheel',
  'resize',
  'mousewheel',
  'animationend',
  'transitionend',
  'touchmove',
];

export const hasIntersectionObserver =
  inBrowser &&
  'IntersectionObserver' in window &&
  'IntersectionObserverEntry' in window &&
  'intersectionRatio' in window.IntersectionObserverEntry.prototype;

export const modeType = {
  event: 'event',
  observer: 'observer',
};

export const call = (fn): void => fn && fn();

export function setImage(
  element: HTMLMediaElement | HTMLIFrameElement | HTMLImageElement | HTMLElement,
  imagePath: string,
): void {
  if (
    element instanceof HTMLIFrameElement ||
    element instanceof HTMLImageElement ||
    element instanceof HTMLMediaElement
  ) {
    element.src = imagePath;
  } else {
    element.style.backgroundImage = `url('${imagePath}')`;
  }
}

export function loadImage(imagePath: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const img = new Image();
    img.src = imagePath;
    img.onload = () => resolve(imagePath);
    img.onerror = (err) => reject(err);
  });
}

export function addCssClassName(
  element: HTMLImageElement | HTMLElement,
  cssClassName: string,
): void {
  if (!element.className.includes(cssClassName)) {
    element.className += ` ${cssClassName}`;
  }
}

export function throttle(fn, threshold = 250, scope?) {
  let last;
  let deferTimer;
  return () => {
    // @ts-ignore
    const context = scope || this;

    const now = +new Date();
    const args = arguments;
    if (last && now < last + threshold) {
      // hold on to it
      clearTimeout(deferTimer);
      deferTimer = setTimeout(() => {
        last = now;
        fn.apply(context, args);
      }, threshold);
    } else {
      last = now;
      fn.apply(context, args);
    }
  };
}

export function debounce(func, wait, immediate?) {
  let timeout;
  let args;
  let context;
  let timestamp;
  let result;

  const later = function later() {
    const last = +new Date() - timestamp;

    if (last < wait && last >= 0) {
      timeout = setTimeout(later, wait - last);
    } else {
      timeout = null;
      if (!immediate) {
        result = func.apply(context, args);
        if (!timeout) {
          context = null;
          args = null;
        }
      }
    }
  };

  return function debounced() {
    // @ts-ignore
    // eslint-disable-next-line @typescript-eslint/no-this-alias
    context = this;
    args = arguments;
    timestamp = +new Date();

    const callNow = immediate && !timeout;
    if (!timeout) {
      timeout = setTimeout(later, wait);
    }

    if (callNow) {
      result = func.apply(context, args);
      context = null;
      args = null;
    }

    return result;
  };
}
