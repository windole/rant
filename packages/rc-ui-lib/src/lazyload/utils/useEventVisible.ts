import { useEffect, useRef, useState } from 'react';
import { debounce as debounceFunc, DEFAULT_EVENTS, throttle as throttleFunc } from '.';
import { EventOptions } from '../PropsType';

// try to handle passive events
let passiveEventSupported = false;
try {
  const opts = Object.defineProperty({}, 'passive', {
    get() {
      passiveEventSupported = true;
    },
  });
  window.addEventListener('test', null, opts);
  // eslint-disable-next-line no-empty
} catch (e) {}

const passiveEvent = passiveEventSupported ? { capture: false, passive: true } : false;

const useEventVisible = (
  eventOptions: EventOptions,
): [React.MutableRefObject<HTMLDivElement | HTMLImageElement>, boolean] => {
  const {
    listenEvents = DEFAULT_EVENTS,
    offset,
    scrollContainer,
    debounce,
    throttle,
  } = eventOptions;
  // 是否可见
  const [visible, setVisible] = useState(false);
  // // 包裹容器
  const containerRef = useRef<HTMLDivElement | HTMLImageElement>(null);

  useEffect(() => {
    // 检查函数
    let checkVisible = () => {
      // 如果容器不存在则不计算
      if (!containerRef.current) return;
      // 获取当前组件位置
      const { top, left, width, height } = containerRef.current?.getBoundingClientRect();
      // 获取滚动容器位置
      const {
        top: parentTop,
        left: parentLeft,
        width: parentWidth,
        height: parentHeight,
      } = scrollContainer.getBoundingClientRect();
      // 计算屏幕高度和宽度
      const windowInnerHeight = window.innerHeight || document.documentElement.clientHeight;
      const windowInnerWidth = window.innerWidth || document.documentElement.clientWidth;
      // 计算滚动容器和屏幕可视区的相交区域
      const intersectionTop = Math.min(parentTop, 0);
      const intersectionLeft = Math.min(parentLeft, 0);

      const intersectionHeight =
        Math.min(windowInnerHeight, parentTop + parentHeight) - intersectionTop;
      const intersectionWidth =
        Math.min(windowInnerWidth, parentLeft + parentWidth) - intersectionLeft;
      // 计算组件距离可视区的高度
      const offsetTop = top - intersectionTop;
      const offsetLeft = left - intersectionLeft;
      // 偏移量计算[top, left]
      const offsets = Array.isArray(offset) ? offset : [offset, offset];
      // 通过上述距离判断组件是否在可见区域
      const isVisible =
        offsetTop - offsets[0] <= intersectionHeight &&
        offsetTop + height + offsets[0] >= 0 &&
        offsetLeft - offsets[1] <= intersectionWidth &&
        offsetLeft + width + offsets[1] >= 0;

      setVisible(isVisible);
      // 如果可见取消当前组件的所有的监听
      if (isVisible) {
        listenEvents.forEach((event) => {
          scrollContainer?.removeEventListener(event, checkVisible);
        });
      }
    };

    // 使用防抖节流增强checkVisible函数，这里优先防抖，其次节流
    if (debounce) {
      // 防抖函数
      checkVisible = debounceFunc(checkVisible, debounce);
    } else if (throttle) {
      // 节流函数
      checkVisible = throttleFunc(checkVisible, throttle);
    }
    // 批量监听事件
    listenEvents.forEach((event) => {
      scrollContainer?.addEventListener(event, checkVisible, passiveEvent);
    });

    // 自动执行一次
    checkVisible();

    return () => {
      listenEvents.forEach((event) => {
        scrollContainer?.removeEventListener(event, checkVisible, passiveEvent);
      });
    };
  }, [scrollContainer]);
  return [containerRef, visible];
};

export default useEventVisible;
