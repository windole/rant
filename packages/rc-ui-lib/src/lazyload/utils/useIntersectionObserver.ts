import { useCallback, useEffect, useState } from 'react';

const useIntersectionObserver = (
  observerOptions: IntersectionObserverInit,
  handleIntersection: IntersectionObserverCallback,
): ((node: Element) => void) => {
  const [element, setElement] = useState<Element>(null);

  const handleElement = useCallback((node: Element) => {
    setElement(node);
  }, []);

  useEffect(() => {
    if (!element) return;
    const observer = new IntersectionObserver(handleIntersection, observerOptions);
    observer.observe(element);
    // eslint-disable-next-line consistent-return
    return () => observer && observer.disconnect && observer.disconnect();
  }, [handleIntersection, observerOptions, element]);
  return handleElement;
};

export default useIntersectionObserver;
