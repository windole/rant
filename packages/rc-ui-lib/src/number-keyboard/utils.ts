import { KeyConfig, NumberKeyboardProps } from './PropsType';

const genBasicKeys = (props: NumberKeyboardProps) => {
  const keys: KeyConfig[] = Array(9)
    .fill('')
    .map((_, i) => ({ text: i + 1 }));

  if (props.randomKeyOrder) {
    keys.sort(() => (Math.random() > 0.5 ? 1 : -1));
  }

  return keys;
};

export const genDefaultKeys = (props: NumberKeyboardProps): KeyConfig[] => [
  ...genBasicKeys(props),
  { text: props.extraKey as string, type: 'extra' },
  { text: 0 },
  {
    text: props.showDeleteKey ? props.deleteButtonText : '',
    type: props.showDeleteKey ? 'delete' : '',
  },
];

export const genCustomKeys = (props: NumberKeyboardProps) => {
  const keys = genBasicKeys(props);
  const { extraKey } = props;
  const extraKeys = Array.isArray(extraKey) ? extraKey : [extraKey];

  if (extraKeys.length === 1) {
    keys.push({ text: 0, wider: true }, { text: extraKeys[0], type: 'extra' });
  } else if (extraKeys.length === 2) {
    keys.push(
      { text: extraKeys[0], type: 'extra' },
      { text: 0 },
      { text: extraKeys[1], type: 'extra' },
    );
  }

  return keys;
};
