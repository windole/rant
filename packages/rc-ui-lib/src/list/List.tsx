import React, { useRef, forwardRef, useImperativeHandle, isValidElement, useContext } from 'react';
import classnames from 'classnames';

import Loading from '../loading';
import { getRect } from '../hooks/get-rect';
import useScrollParent from '../hooks/use-scroll-parent';
import useEventListener from '../hooks/use-event-listener';

import { isHidden } from '../utils';
import { ListInstance, ListProps } from './PropsType';
import { useSetState, useUpdateEffect } from '../hooks';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';

const List = forwardRef<ListInstance, ListProps>((props, ref) => {
  const { prefixCls, createNamespace } = useContext(ConfigProviderContext);
  const [bem] = createNamespace('list', prefixCls);
  const [state, updateState] = useSetState({
    loading: props.loading,
    error: props.error,
  });

  const errorRecent = useRef(props.error);

  const root = useRef<HTMLDivElement>();
  const scrollParent = useRef(null);
  const placeholder = useRef<HTMLDivElement>();

  scrollParent.current = useScrollParent(root);

  // 判断是否需要加载
  const check = async () => {
    if (!props.onLoad) return;
    if (state.loading || props.finished || errorRecent.current) {
      return;
    }
    const { offset, direction } = props;
    const scrollParentRect = getRect(scrollParent.current);

    if (!scrollParentRect.height || isHidden(root.current)) {
      return;
    }

    let isReachEdge = false;
    const placeholderRect = getRect(placeholder.current);
    if (direction === 'up') {
      isReachEdge = scrollParentRect.top - placeholderRect.top <= offset;
    } else {
      isReachEdge = placeholderRect.bottom - scrollParentRect.bottom <= offset;
    }
    if (isReachEdge) {
      try {
        updateState({ loading: true });
        if (props.onLoad) await props.onLoad();
        updateState({ loading: false });
      } catch (error) {
        // eslint-disable-next-line no-console
        console.warn('onLoad error:', error);
        updateState({ loading: false, error: true });
        errorRecent.current = true;
      }
    }
  };

  const renderFinishedText = () => {
    if (props.finished && props.finishedText) {
      return <div className={classnames(bem('finished-text'))}>{props.finishedText}</div>;
    }
    return null;
  };

  const clickErrorText = () => {
    updateState({ error: false });
    errorRecent.current = false;
    check();
  };

  const renderErrorText = () => {
    if (state.error && props.errorText) {
      return (
        <div className={classnames(bem('error-text'))} onClick={clickErrorText}>
          {props.errorText}
        </div>
      );
    }
    return null;
  };

  const renderLoading = () => {
    if (state.loading && !props.finished) {
      return (
        <div className={classnames(bem('loading'))}>
          {isValidElement(props.loadingText) ? (
            props.loadingText
          ) : (
            <Loading className={classnames(bem('loading-icon'))} size={16}>
              {props.loadingText}
            </Loading>
          )}
        </div>
      );
    }
    return null;
  };

  useUpdateEffect(() => {
    if (props.autoCheck) {
      check();
    }
  }, [state.loading, props.finished, props.error]);

  useUpdateEffect(() => {
    updateState({ loading: props.loading, error: props.error });
    errorRecent.current = props.error;
  }, [props.loading, props.error]);

  useUpdateEffect(() => {
    if (scrollParent.current && props.immediateCheck) {
      check();
    }
  }, [scrollParent.current]);

  useEventListener('scroll', check, {
    target: scrollParent.current,
    depends: [state.loading, props.finished, state.error],
  });

  useImperativeHandle(ref, () => ({
    check,
    state,
  }));

  const Placeholder = <div ref={placeholder} className={classnames(bem('placeholder'))} />;

  return (
    <div ref={root} role="feed" className={classnames(bem())} aria-busy={state.loading}>
      {props.direction === 'down' ? props.children : Placeholder}
      {renderLoading()}
      {renderFinishedText()}
      {renderErrorText()}
      {props.direction === 'up' ? props.children : Placeholder}
    </div>
  );
});

List.defaultProps = {
  offset: 300,
  direction: 'down',
  immediateCheck: true,
  autoCheck: true,
  loadingText: '加载中...',
  finishedText: '没有更多了',
};

export default List;
