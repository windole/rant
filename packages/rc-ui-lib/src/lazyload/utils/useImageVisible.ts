import { useCallback, useState } from 'react';
import { addCssClassName, loadImage, setImage } from '.';
import { IntersectionObserverEntryType, ObserverOptions } from '../PropsType';
import useIntersectionObserver from './useIntersectionObserver';

const defaultObserverOptions = {
  threshold: [0, 1],
};

const useImageVisible = (
  observerOptions: ObserverOptions = defaultObserverOptions,
  errorImage: string,
): [(node: Element) => void, boolean] => {
  const [isLoaded, setIsLoaded] = useState(false);

  const handleIntersection = useCallback(
    (entries: IntersectionObserverEntryType[]) => {
      entries
        .filter((entry) => entry.isIntersecting)
        .forEach((entry) => {
          const target = entry.target as HTMLImageElement | HTMLElement;
          const url = target.dataset.src || '';
          loadImage(url)
            .catch(() => {
              setImage(target, errorImage);
              addCssClassName(target, 'rc-lazyload__error');
              target.setAttribute('lazy', 'error');
            })
            .then((imagePath: string) => {
              setImage(target, imagePath);
              setIsLoaded(true);
              addCssClassName(target, 'rc-lazyload__loaded');
              target.setAttribute('lazy', 'loaded');
              target.removeAttribute('data-src');
            });
        });
    },
    [setIsLoaded, isLoaded],
  );
  const ref = useIntersectionObserver(observerOptions, handleIntersection);
  return [ref, isLoaded];
};

export default useImageVisible;
