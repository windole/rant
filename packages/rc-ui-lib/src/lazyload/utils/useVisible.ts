import { useCallback, useEffect, useState } from 'react';
import { addCssClassName } from '.';
import useIntersectionObserver from './useIntersectionObserver';

const defaultObserverOptions = {
  threshold: [0, 1],
};

const useVisible = (
  initialVisible: boolean,
  observerOptions = defaultObserverOptions,
): [(node: Element) => void, boolean] => {
  const [visible, setVisible] = useState(initialVisible);

  const forceVisible = () => {
    setVisible(true);
  };

  useEffect(() => {
    if (initialVisible) {
      forceVisible();
    }
  }, [initialVisible]);

  const handleIntersection = useCallback(
    (entries) => {
      entries
        .filter((entry) => entry.isIntersecting)
        .forEach((entry) => {
          const target = entry.target as HTMLElement;
          addCssClassName(target, 'rc-lazyload__loaded');
          setVisible(true);
        });
    },
    [setVisible, visible],
  );
  const ref = useIntersectionObserver(observerOptions, handleIntersection);
  return [ref, visible];
};

export default useVisible;
