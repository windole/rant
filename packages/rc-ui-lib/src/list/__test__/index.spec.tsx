import React from 'react';
import { fireEvent, render, waitFor, act } from '@testing-library/react';
import List from '..';
import Cell from '../../cell';
import { sleep } from '../../../tests/utils';

describe('List', () => {
  const mockRect = (el: Element, rect: Partial<DOMRect>) => {
    Object.defineProperty(el, 'getBoundingClientRect', {
      configurable: true,
      get: () => {
        return () => ({
          top: 0,
          left: 0,
          right: 0,
          bottom: 0,
          width: 0,
          height: 0,
          ...rect,
        });
      },
    });
  };

  beforeEach(() => {
    // mocking the offsetParent
    Object.defineProperty(HTMLElement.prototype, 'offsetParent', {
      get() {
        return {};
      },
    });
  });
  it('should emit load event when reaching bottom', async () => {
    const onLoad = jest.fn();

    await act(async () => {
      const { rerender } = render(<List onLoad={onLoad} />);

      await sleep(100);
      rerender(<List finished onLoad={onLoad} />);
    });

    expect(onLoad).toHaveBeenCalled();
  });

  it('should render loading tip correctly and do not trigger onLoad', () => {
    const onLoad = jest.fn();
    const { container } = render(<List loading onLoad={onLoad} />);

    expect(container).toMatchSnapshot();

    expect(onLoad).not.toBeCalled();
  });

  it('should render finish tip correctly and do not trigger onLoad', () => {
    const onLoad = jest.fn();
    const { container } = render(<List finished finishedText="finished" onLoad={onLoad} />);

    expect(container).toMatchSnapshot();
    expect(onLoad).not.toBeCalled();
  });

  it('should render error slots correctly and do not trigger onLoad', async () => {
    const onLoad = jest.fn();
    const { getByText, container } = render(<List error errorText="errorTip!" />);

    expect(container).toMatchSnapshot();
    expect(onLoad).not.toBeCalled();

    const errorTextBox = getByText('errorTip!');

    act(() => {
      fireEvent.click(errorTextBox);
    });
    expect(onLoad).toHaveBeenCalledTimes(0);
  });

  it('should render correctly when set direction', async () => {
    const onLoad = jest.fn();
    const { container, rerender } = render(<List direction="up" onLoad={onLoad} />);

    await act(async () => {
      rerender(<List direction="up" loading onLoad={onLoad} />);
      await sleep(100);
      rerender(<List direction="up" onLoad={onLoad} />);
      await sleep(100);
      rerender(<List direction="up" finished onLoad={onLoad} />);
    });

    expect(container).toMatchSnapshot();

    expect(onLoad).toHaveBeenCalled();
  });

  it('should trigger load when scroll', async () => {
    Object.defineProperty(window.HTMLElement.prototype, 'clientHeight', {
      value: 640,
    });

    const onLoad = jest.fn();

    const { container, rerender } = render(
      <List onLoad={onLoad} offset={100} immediateCheck={false}>
        {new Array(30).fill(0).map((_, i) => (
          <Cell key={String(i)} title={i} />
        ))}
      </List>,
    );

    const listContainer = container.querySelector('.rc-list');
    const placeholderContainer = container.querySelector('.rc-list__placeholder');

    Object.defineProperty(listContainer, 'offsetParent', {
      configurable: true,
      get: () => document.body,
    });
    mockRect(listContainer, {
      top: 0,
      bottom: 750,
    });

    mockRect(placeholderContainer, {
      top: 600,
      bottom: 780,
    });

    expect(container).toMatchSnapshot();

    fireEvent.scroll(window, { target: { pageYOffset: 50 } });

    rerender(
      <List finished offset={100} immediateCheck={false}>
        {new Array(30).fill(0).map((_, i) => (
          <Cell key={String(i)} title={i} />
        ))}
      </List>,
    );

    await waitFor(() => {
      expect(onLoad).toHaveBeenCalled();
    });
  });
});
