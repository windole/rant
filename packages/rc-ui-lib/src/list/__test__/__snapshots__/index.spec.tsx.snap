// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render correctly when set direction 1`] = `
<div>
  <div
    aria-busy="false"
    class="rc-list"
    role="feed"
  >
    <div
      class="rc-list__placeholder"
    />
    <div
      class="rc-list__finished-text"
    >
      没有更多了
    </div>
  </div>
</div>
`;

exports[`List should render error slots correctly and do not trigger onLoad 1`] = `
<div>
  <div
    class="rc-list"
    role="feed"
  >
    <div
      class="rc-list__error-text"
    >
      errorTip!
    </div>
    <div
      class="rc-list__placeholder"
    />
  </div>
</div>
`;

exports[`List should render finish tip correctly and do not trigger onLoad 1`] = `
<div>
  <div
    class="rc-list"
    role="feed"
  >
    <div
      class="rc-list__finished-text"
    >
      finished
    </div>
    <div
      class="rc-list__placeholder"
    />
  </div>
</div>
`;

exports[`List should render loading tip correctly and do not trigger onLoad 1`] = `
<div>
  <div
    aria-busy="true"
    class="rc-list"
    role="feed"
  >
    <div
      class="rc-list__loading"
    >
      <div
        class="rc-list__loading-icon rc-loading rc-loading--circular"
      >
        <span
          class="rc-loading__spinner rc-loading__spinner--circular"
          style="width: 16px; height: 16px;"
        >
          <svg
            class="rc-loading__circular"
            viewBox="25 25 50 50"
          >
            <circle
              cx="50"
              cy="50"
              fill="none"
              r="20"
            />
          </svg>
        </span>
        <span
          class="rc-loading__text"
        >
          加载中...
        </span>
      </div>
    </div>
    <div
      class="rc-list__placeholder"
    />
  </div>
</div>
`;

exports[`List should trigger load when scroll 1`] = `
<div>
  <div
    class="rc-list"
    role="feed"
  >
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        0
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        1
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        2
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        3
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        4
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        5
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        6
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        7
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        8
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        9
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        10
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        11
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        12
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        13
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        14
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        15
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        16
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        17
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        18
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        19
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        20
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        21
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        22
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        23
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        24
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        25
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        26
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        27
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        28
      </div>
    </div>
    <div
      class="rc-cell"
    >
      <div
        class="rc-cell__title"
      >
        29
      </div>
    </div>
    <div
      class="rc-list__placeholder"
    />
  </div>
</div>
`;
